### PR TITLE
Adding new feature for save image digest information [WIP]

### LINKED ISSUE #57: Add immutable identifier over a Docker image
SAPS currently uses the tag name and repository of the 3 phases of a task's execution flow as information of what version the Arrebol service will use when processing that task. But if the version of the Docker image changes in the future it is not possible to recover the state of the version that was processed in that particular phase, so it is necessary to add 3 new information about the immutable identifier of the Docker image in Catalog and use it to run in Arrebol.

### PR DESCRIPTION
**Pull request by thiagoyeds**
Fix #57

### CODE REVIEW

File: scripts/get_digest.sh [hunk 1]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+
```
Comment (thiagomanel) [inline scripts/get_digest.sh]: add some doc here. explain the purpose of this script, the parameters and return values

Comment (thiagomanel) [reply scripts/get_digest.sh]: also, remove the ".sh" termination. an executable script usually does not use .sh termination

File: scripts/get_digest.sh [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+IMAGE=$1
+TAG=$2
+
+CURL=`which curl`
```
Comment (thiagomanel) [inline scripts/get_digest.sh]: why using `which` here?

File: scripts/get_digest.sh [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+JQ=`which jq`
+SED=`which sed`
+
+HOST_REGISTRY="https://hub.docker.com/v2"
+PATH="repositories/$IMAGE/tags"
+
+DIGEST=`$CURL -f --silent --header "Accept: application/json" $HOST_REGISTRY/$PATH | $JQ ".results[] | select(.name == \"$TAG\") | .images[].digest" | $SED -e 's/\"//g'`
+
```
Comment (thiagomanel) [inline scripts/get_digest.sh]: I would add some testing on the `DIGEST` string. for example, it must have a specific length right?

File: src/main/java/org/fogbowcloud/saps/engine/core/database/ImageDataStore.java [hunk 4]
```diff
@@ -25,7 +25,8 @@
 	public String DATASTORE_PORT = "datastore_port";
 
 	SapsImage addImageTask(String taskId, String dataset, String region, Date date, int priority, String user,
-			String inputGathering, String inputPreprocessing, String algorithmExecution) throws SQLException;
+			String inputdownloadingPhaseTag, String digestInputdownloading, String preprocessingPhaseTag,
```
Comment (thiagomanel) [inline src/main/java/org/fogbowcloud/saps/engine/core/database/ImageDataStore.java:28]: your modifications look good to me. However, I think you should take the opportunity to refactor the code (you can choose to modify in this PR or to open another one, with refactor-only modifications).

1. delete all unused methods from this interface
2. This `ImageDataStore` is our `Catalog` right? If so, change the name.
3. do we need two `addImageTask` method? Maybe, we can keep the one that receives `SapsImage` parameter

Comment (thiagomanel) [reply src/main/java/org/fogbowcloud/saps/engine/core/database/ImageDataStore.java:28]: also, there is no need to use `public` in the fields declared in an interface

File: src/main/java/org/fogbowcloud/saps/engine/core/dispatcher/SubmissionDispatcher.java [hunk 5]
```diff
@@ -238,13 +241,18 @@ else if (rowAux.length() == 2)
 	 * @param inputdownloadingPhaseTag inputdownloading phase tag
 	 * @param preprocessingPhaseTag    preprocessing phase tag
 	 * @param processingPhaseTag       processing phase tag
+	 * @param digestProcessing
+	 * @param processingPhaseTag2
```
Comment (thiagomanel) [inline src/main/java/org/fogbowcloud/saps/engine/core/dispatcher/SubmissionDispatcher.java:245]: what is a `processingPhaseTag2`. In fact, we need to add *good* docs related to this tag and digest stuff. I mean, with some examples. This is non-trivial stuff (so, need proper doc).

As I commented before, the code you added looks good to me but we need to refactor legacy code (you can make it in another PR):
1. remove silly documentation (e.g. "This function adds new user." in a method called `addNewUserInCatalog` does not add much, so remove the setence)
2. remove dead code
3. overall, we have important code in classes name *Util (e.g. CatalogUtils) and ancillary code in the main classes (e.g. formatPathRow)

File: src/main/java/org/fogbowcloud/saps/engine/core/dto/TaskRequestDTO.java [hunk 6]
```diff
@@ -10,14 +10,14 @@
...
     private static final long serialVersionUID = 1L;
     
-    private String id;
+    private String label;
     private Map<String, String> requirements;
     private List<String> commands;
     private Map<String, String> metadata;
 
-    public TaskRequestDTO(String id, Map<String, String> requirements, List<String> commands,
+    public TaskRequestDTO(String label, Map<String, String> requirements, List<String> commands,
```
Comment (thiagomanel) [inline src/main/java/org/fogbowcloud/saps/engine/core/dto/TaskRequestDTO.java:18]: this is non-trival. what is a `label`?. we should make the life of new devs easier.

File: src/main/java/org/fogbowcloud/saps/engine/core/model/SapsImage.java [hunk 7]
```diff
@@ -33,16 +33,20 @@
...
+	private String digestProcessing;
 	private Timestamp creationTime;
 	private Timestamp updateTime;
 	private String status;
 	private String error;
 
-	public SapsImage(String taskId, String dataset, String region, Date imageDate, ImageTaskState state, String arrebolJobId,
-					 String federationMember, int priority, String user, String inputdownloadingTag, String preprocessingTag,
-					 String processingTag, Timestamp creationTime, Timestamp updateTime, String status, String error) {
+	public SapsImage(String taskId, String dataset, String region, Date imageDate, ImageTaskState state,
```
Comment (thiagomanel) [inline src/main/java/org/fogbowcloud/saps/engine/core/model/SapsImage.java:46]: we need proper doc in this constructor (you can make in a refactor PR, latter)
do we need all these set methods here?
I don't link this SapsImage name. You should discuss a better option

**Review by thiagomanel — CHANGES_REQUESTED** [review]
overall, LGTM.
two major remarks:
1. add some doc related to digests
2. we need some refactoring (you can add commits to this PR or create a new PR to add refactor-only code, 
I think the second option is better).

File: scripts/get_digest.sh (hunk 1 above)
Comment (thiagoyeds) [reply scripts/get_digest.sh]: done

File: scripts/get_digest.sh (hunk 2 above)
Comment (thiagoyeds) [reply scripts/get_digest.sh]: For some reason I was having trouble using the commands directly in the script, something like:
```
curl: command not found

sed: command not found

jq: command not found
```
I found it safer to do it that way.

File: scripts/get_digest.sh (hunk 3 above)
Comment (thiagoyeds) [reply scripts/get_digest.sh]: Test added

File: src/main/java/org/fogbowcloud/saps/engine/core/database/ImageDataStore.java (hunk 4 above)
Comment (thiagoyeds) [reply src/main/java/org/fogbowcloud/saps/engine/core/database/ImageDataStore.java:28]: I created an issue (#73 ) reporting on.

File: src/main/java/org/fogbowcloud/saps/engine/core/dispatcher/SubmissionDispatcher.java (hunk 5 above)
Comment (thiagoyeds) [reply src/main/java/org/fogbowcloud/saps/engine/core/dispatcher/SubmissionDispatcher.java:245]: I created an other issue (#74) reporting on.

**Comment by thiagoyeds** [comment]
> overall, LGTM.
> two major remarks:
> 
> 1. add some doc related to digests
> 2. we need some refactoring (you can add commits to this PR or create a new PR to add refactor-only code,
>    I think the second option is better).

@thiagomanel, would this document reporting the digest be a markdown in the _/docs_ folder?

File: src/main/java/org/fogbowcloud/saps/engine/core/dto/TaskRequestDTO.java (hunk 6 above)
Comment (thiagoyeds) [reply src/main/java/org/fogbowcloud/saps/engine/core/dto/TaskRequestDTO.java:18]: I created a problem (#75) to work on refactoring the code.

File: src/main/java/org/fogbowcloud/saps/engine/core/model/SapsImage.java (hunk 7 above)
Comment (thiagoyeds) [reply src/main/java/org/fogbowcloud/saps/engine/core/model/SapsImage.java:46]: I created a problem (#75) to work on refactoring the code.

**Review by thiagomanel — APPROVED** [review]
ok, lgtm.  thanks @ThiagoWhispher

**Pull request merged**